### PR TITLE
Fieldsolver dt2, boundary, and moment fixes

### PR DESCRIPTION
**Pull request by markusbattarbee**
Several fixes:

1) accelerates inflowing Maxwellian cells so that (in case of low v-space resolution) they are in sync with accelerated inner cells, removing the inflow artefact which was being built at the boundary.

2) Applies boundary conditions and re-calculates moments at end of initialization. This fixes some corner blips in fields.

3) Corrects the ordering of use of moments and moments_dt2 in RK2 fieldsolver. I think this approach is correct and better matches classical 2nd-order-accurate schemes. This includes changes to ghost updates etc to match the requirements.

4) Perhaps controversial: perB_dt2 (at t=t'+dt/2) is now updated as a dt-length step from the previous round perB_dt2 (at t=t'-dt/2), using E at t=t'. Previously, it was built on top of perB at t=t'. This method is more of a proper 2nd-order-accurate method, but I am not yet completely convinced it doesn't contain the risk of diverging B-grids at the full and intermediate steps.

5) re-activate Hall term derivatives communication for phase RK_ORDER2_STEP2 when EgradPe-term is active. This was overlooked in previous approach.

### CODE REVIEW

**Comment by markusbattarbee** [comment]
Having run dispersion tests on a local cluster, I can't show any significant difference between these versions. Next step is to see how this version fares for global runs where FS-dt falls.

File: fieldsolver/ldz_magnetic_field.cpp [hunk 1]
```diff
@@ -86,12 +86,20 @@ void propagateMagneticField(
...
             perBDt2Grid0->at(fsgrids::bfield::PERBX) = perBGrid0->at(fsgrids::bfield::PERBX) + 0.5*dt*(1.0/dz*(EGrid2->at(fsgrids::efield::EY) - EGrid0->at(fsgrids::efield::EY)) + 1.0/dy*(EGrid0->at(fsgrids::efield::EZ) - EGrid1->at(fsgrids::efield::EZ)));
             break;
-            
+
+         case RK_ORDER2_STEP1SS:
+            perBDt2Grid0 = perBDt2Grid.get(i,j,k);
+            EGrid0 = EGrid.get(i,j,k);
+            EGrid1 = EGrid.get(i,j+1,k);
+            EGrid2 = EGrid.get(i,j,k+1);
+            perBDt2Grid0->at(fsgrids::bfield::PERBX) += dt/dz*(EGrid2->at(fsgrids::efield::EY) - EGrid0->at(fsgrids::efield::EY)) + dt/dy*(EGrid0->at(fsgrids::efield::EZ) - EGrid1->at(fsgrids::efield::EZ));
```
Comment (ykempf) [inline fieldsolver/ldz_magnetic_field.cpp:95]: This (and the next two similar lines) should be `=` and not `+=` I think. Compare with line 87.

Comment (markusbattarbee) [reply fieldsolver/ldz_magnetic_field.cpp:95]: No, I think this is as intended. Line 87, for the first of a group of substeps, evaluates perB_dt2 as a half-step ahead from perB.
This line evaluates the n>1 substep perB_dt2 as a full-timestep advance from the perB_dt2 of the previous substep.

I'll check out what's happening with boundary conditions though as that's where the blowup was located.

**Comment by markusbattarbee** [comment]
So apparently these tests failed due to always creating a newFSgrid object and though deleting it after the fieldsolver portion, being unable to initiate new MPI communicators:
```
aborting job:
Fatal error in PMPI_Cart_create: Other MPI error, error stack:
PMPI_Cart_create(302)...............: MPI_Cart_create(MPI_COMM_WORLD, ndims=3, dims=0x7fff890c9d00, periods=0x7fff890c9af0, reorder=0, comm_cart=0x7fff890c9c50) failed
MPIR_Cart_create_impl(179)..........: 
MPIR_Cart_create(112)...............: 
MPII_Comm_copy(544).................: 
MPICH ERROR [Rank 96] [job id 317493.1] [Sun Dec  5 02:48:54 2021] [nid001402] - Abort(1008826895) (rank 96 in comm 0): Fatal error in PMPI_Cart_create: Other MPI error, error stack:
PMPI_Cart_create(302)...............: MPI_Cart_create(MPI_COMM_WORLD, ndims=3, dims=0x7ffd47988300, periods=0x7ffd479880f0, reorder=0, comm_cart=0x7ffd47988250) failed
MPIR_Cart_create_impl(179)..........: 
MPIR_Cart_create(112)...............: 
MPII_Comm_copy(544).................: 
MPIR_Get_contextid_sparse_group(679): Too many communicators (0/2048 free on this process; ignore_id=0)
```
Does not the implicit FSgrid descructor (due to grids going out of scope) not free up the communicator?

**Comment by markusbattarbee** [comment]
Nope, even adding the finalize calls did not solve the issue.
```
aborting job:
Fatal error in PMPI_Cart_create: Other MPI error, error stack:
PMPI_Cart_create(302)...............: MPI_Cart_create(MPI_COMM_WORLD, ndims=3, dims=0x7fff65478730, periods=0x7fff65477a30, reorder=0, comm_cart=0x7fff65478680) failed
MPIR_Cart_create_impl(179)..........: 
MPIR_Cart_create(112)...............: 
MPII_Comm_copy(544).................: 
MPICH ERROR [Rank 195] [job id 610927.1] [Mon Dec 20 10:28:02 2021] [nid001597] - Abort(539064847) (rank 195 in comm 0): Fatal error in PMPI_Cart_create: Other MPI error, error stack:
PMPI_Cart_create(302)...............: MPI_Cart_create(MPI_COMM_WORLD, ndims=3, dims=0x7ffe96defbf0, periods=0x7ffe96deeef0, reorder=0, comm_cart=0x7ffe96defb40) failed
MPIR_Cart_create_impl(179)..........: 
MPIR_Cart_create(112)...............: 
MPII_Comm_copy(544).................: 
MPIR_Get_contextid_sparse_group(679): Too many communicators (0/2048 free on this process; ignore_id=0)

**Comment by kstppd** [comment]
Ok so this is weird if it is still failing after calling finalize().

**Comment by markusbattarbee** [comment]
The following three sub-parts were already merged in #615:

1.  accelerates inflowing Maxwellian cells so that (in case of low v-space resolution) they are in sync with accelerated inner cells, removing the inflow artefact which was being built at the boundary.

2.  Applies boundary conditions and re-calculates moments at end of initialization. This fixes some corner blips in fields.

5.  re-activate Hall term derivatives communication for phase RK_ORDER2_STEP2 when EgradPe-term is active. This was overlooked in previous approach.
